File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for(String s : strs)
        {
            int[] chars = new int[26];
            for(char c : s.toCharArray())
            {
                chars[c - 'a']++;
            }
            String key = Arrays.toString(chars);
            List<String> existingValues = map.getOrDefault(key,new ArrayList<>());
            existingValues.add(s);
            map.put(key, existingValues);
        }
        return new ArrayList<>(map.values());
    }
}

File: 0049-group-anagrams/0049-group-anagrams.py
from collections import defaultdict
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        buckets = defaultdict(list)
        for string in strs:
            characters = [0] * 26
            for char in string:
                characters[ord('z') - ord(char)] += 1
            buckets[tuple(characters)].append(string)
        return list(buckets.values())



